walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,55 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> eventTimesOnlyMandatory = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> eventTimesWithOptional = new ArrayList<TimeRange>();
+    for (Event event : events) {
+	  Set<String> intersection = new HashSet<String>(request.getAttendees());
+	  intersection.retainAll(event.getAttendees());
+      if (intersection.size() > 0) {
+        eventTimesOnlyMandatory.add(event.getWhen());
+        eventTimesWithOptional.add(event.getWhen());
+      } else {
+        intersection = new HashSet<String>(request.getOptionalAttendees());
+        intersection.retainAll(event.getAttendees());
+        if(intersection.size() > 0) {
+          eventTimesWithOptional.add(event.getWhen());
+        }
+      }
+    }
+    Collection<TimeRange> options =  findFreeTimes(eventTimesWithOptional, request.getDuration());
+    if (options.size() > 0 || request.getAttendees().size() == 0 || request.getOptionalAttendees().size() == 0) {
```

Comment:
you can move `request.getAttendees().size() == 0 || request.getOptionalAttendees().size() == 0` to the top of the function and skip all the processing since this condition doesn't depend on the results

Reply:
I agree that the condition isn't dependent on the processing, but isn't the body of the if statement dependent on the processing?

Reply:
Oh you're right, my bad!